File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void helper(vector<int> &candidates, vector<vector<int>> &ans, vector<int> &v, int target, int index){
        if(target==0){
            ans.push_back(v);
            return;
        }
        if(target<0){
            return;
        }

        for(int i=index; i<candidates.size(); i++){
            v.push_back(candidates[i]);
            helper(candidates, ans, v, target-candidates[i], i);
            v.pop_back();
        }
        return;
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>v;
        helper(candidates, ans, v, target, 0);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtracking(0, current, candidates, target, result);
        return result;
    }
    public void backtracking(int index, List<Integer> current, int[] candidates, int target, List<List<Integer>> result){
        if(target==0){
            result.add(new ArrayList<Integer>(current));
            return;
        }
        else if(target<0){
            return;
        }
        for(int i=index;i<candidates.length;i++){
            current.add(candidates[i]);
            backtracking(i, current, candidates, target-candidates[i], result);
            current.remove(current.size()-1);
        }
    }
}